presto-main/src/main/java/com/facebook/presto/sql/planner/PlannerUtils.java

```diff
@@ -87,4 +123,155 @@ public static VariableReferenceExpression toVariableReference(Expression express
         checkArgument(expression instanceof SymbolReference);
         return variable(getSourceLocation(expression), ((SymbolReference) expression).getName(), types.get(expression));
     }
+
+    public static Type createMapType(FunctionAndTypeManager functionAndTypeManager, Type keyType, Type valueType)
+    {
+        MethodHandle keyNativeEquals = functionAndTypeManager.getJavaScalarFunctionImplementation(functionAndTypeManager.resolveOperator(OperatorType.EQUAL, fromTypes(keyType, keyType))).getMethodHandle();
+        MethodHandle keyNativeHashCode = functionAndTypeManager.getJavaScalarFunctionImplementation(functionAndTypeManager.resolveOperator(OperatorType.HASH_CODE, fromTypes(keyType))).getMethodHandle();
+        return new MapType(keyType, valueType, keyNativeEquals, keyNativeHashCode);
+    }
+
+    public static RowExpression orNullHashCode(RowExpression expression)
+    {
+        checkArgument(BIGINT.equals(expression.getType()), "expression should be BIGINT type");
+        return new SpecialFormExpression(expression.getSourceLocation(), SpecialFormExpression.Form.COALESCE, BIGINT, expression, constant(NULL_HASH_CODE, BIGINT));
+    }
+
+    public static Optional<RowExpression> getHashExpression(FunctionAndTypeManager functionAndTypeManager, List<VariableReferenceExpression> variables)
+    {
+        if (variables.isEmpty()) {
+            return Optional.empty();
+        }
+
+        RowExpression result = constant(INITIAL_HASH_VALUE, BIGINT);
+        for (VariableReferenceExpression variable : variables) {
+            RowExpression hashField = call(functionAndTypeManager, HASH_CODE, BIGINT, variable);
+            hashField = orNullHashCode(hashField);
+            result = call(functionAndTypeManager, "combine_hash", BIGINT, result, hashField);
+        }
+        return Optional.of(result);
+    }
+
+    public static PlanNode projectExpressions(PlanNode source, PlanNodeIdAllocator planNodeIdAllocator, VariableAllocator variableAllocator, List<? extends RowExpression> expressions)
```

Comment:
thanks for adding these. I think I have some implementation in MV which can use these.